protocol-testing/src/rpc.rs

```diff
@@ -81,6 +121,314 @@ impl RPCProvider {
+            let storage_slot = self.calculate_executor_storage_slot(executor_address);
+
+            // Double check that this executor is actually approved.
+            // TODO this can be simplified to just set the value to 1 every time.
+            //  This explicit check was just for debug purposes to verify our storage slot calculation.
+            match self
+                .get_storage_at(contract_address, storage_slot)
+                .await
+            {
+                Ok(value) => {
+                    if !value.is_zero() {
+                        info!(
+                            "Found executor approval for {} ({:?}): 0x{}",
+                            protocol_name,
+                            executor_address,
+                            hex::encode(value)
+                        );
+                        state_override = state_override.with_state_diff(
+                            alloy::primitives::Bytes::from(storage_slot.to_vec()),
+                            alloy::primitives::Bytes::from(value.to_vec()),
+                        );
+                    } else {
+                        info!(
+                            "Executor {} ({:?}) is not approved (value is zero)",
+                            protocol_name, executor_address
+                        );
+                    }
+                }
+                Err(e) => {
+                    info!(
+                        "Failed to fetch executor approval for {} ({:?}): {}",
+                        protocol_name, executor_address, e
+                    );
+                }
+            }
+        }
+
+        info!("Completed storage fetch for contract {:?}", contract_address);
+        Ok(state_override)
+    }
+
+    /// Calculate storage slot for Solidity mapping.
+    ///
+    /// The solidity code:
+    /// keccak256(abi.encodePacked(bytes32(key), bytes32(slot)))
+    pub fn calculate_executor_storage_slot(&self, key: Address) -> FixedBytes<32> {
+        // Convert key (20 bytes) to 32-byte left-padded array (uint256)
+        let mut key_bytes = [0u8; 32];
+        key_bytes[12..].copy_from_slice(key.as_slice());
+
+        // The base of the executor storage slot is 1, since there is only one
+        // variable that is initialized before it (which is _roles in AccessControl.sol).
+        // In this case, _roles gets slot 0.
+        // The slots are given in order to the parent contracts' variables first and foremost.
+        let slot = U256::from(1);
+
+        // Convert U256 slot to 32-byte big-endian array
+        let slot_bytes = slot.to_be_bytes::<32>();
+
+        // Concatenate key_bytes + slot_bytes, then keccak hash
+        let mut buf = [0u8; 64];
+        buf[..32].copy_from_slice(&key_bytes);
+        buf[32..].copy_from_slice(&slot_bytes);
+        keccak256(&buf)
+    }
+
+    // TODO remove this. This was useful for initial debugging, but is ultimately useless for
+    //  simulations on past blocks.
```

Comment:
why is it useless?

Reply:
because it doesn't let you simulate on past blocks